### PR TITLE
Matching performance fix

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
This fixes a performance regression MBM introduced for a certain edge case that happens to be very common in rspace, especially when running the 'wide.rho' test.

cc @pyrocto

### JIRA issue:
n/a

### Checklist
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### CODE REVIEW

**Comment by goral09** [comment]
Could you add a comment to the `else if` case that explains it?

**Comment by ArturGajowy** [comment]
> Could you add a comment to the else if case that explains it?

I'd like to refactor this part a bit once the dust settles, so I'll refrain for now :)

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala [hunk 1]
```diff
@@ -102,6 +106,17 @@ class VarMatcherSpec extends FlatSpec with Matchers {
     assertSpatialMatch(target, pattern, Some(Map[Int, Par](0 -> GInt(7), 1 -> GInt(8))))
   }
 
+  "Matching huge list of targets with no patterns and a reminder" should "better be quick" in {
+    //This is a very common case in rspace that can be handled in linear time, yet was quadratic for a short while
+    val target: Par = Par(exprs = Seq.fill(1000)(GInt(1): Expr))
+    val pattern: Par = EVar(FreeVar(0))
+
+    import org.scalatest.time.SpanSugar._
+    failAfter(5 seconds) {
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: Doest this take ~3s on your laptop?

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala [hunk 2]
```diff
@@ -102,6 +106,17 @@ class VarMatcherSpec extends FlatSpec with Matchers {
     assertSpatialMatch(target, pattern, Some(Map[Int, Par](0 -> GInt(7), 1 -> GInt(8))))
   }
 
+  "Matching huge list of targets with no patterns and a reminder" should "better be quick" in {
```
Comment (lukasz-golebiewski) [inline rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: *remainder

File: rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala (hunk 1 above)
Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: less than 1 s, and more than 15 s with the quadratic regression

Comment (lukasz-golebiewski) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: OK. I fear this might fail in CI from time to time

Comment (ArturGajowy) [reply rholang/src/test/scala/coop/rchain/rholang/interpreter/matcher/MatchTest.scala]: So do I, but we can then bump the timeout to 10s or so - that should still differentiate a quadratic regression reliably
